crates/bevy_render/src/view/visibility/mod.rs

```diff
@@ -306,83 +296,108 @@ pub fn update_frusta<T: Component + CameraProjection + Send + Sync + 'static>(
 }
 
 fn visibility_propagate_system(
-    mut root_query: Query<
-        (
-            Option<&Children>,
-            &Visibility,
-            &mut ComputedVisibility,
-            Entity,
-        ),
-        Without<Parent>,
+    changed: Query<
+        (Entity, &Visibility, Option<&Parent>, Option<&Children>),
+        (With<InheritedVisibility>, Changed<Visibility>),
     >,
-    mut visibility_query: Query<(&Visibility, &mut ComputedVisibility, &Parent)>,
-    children_query: Query<&Children, (With<Parent>, With<Visibility>, With<ComputedVisibility>)>,
+    mut visibility_query: Query<(&Visibility, &mut InheritedVisibility)>,
+    children_query: Query<&Children, (With<Visibility>, With<InheritedVisibility>)>,
 ) {
-    for (children, visibility, mut computed_visibility, entity) in root_query.iter_mut() {
-        // reset "view" visibility here ... if this entity should be drawn a future system should set this to true
-        computed_visibility
-            .reset(visibility == Visibility::Inherited || visibility == Visibility::Visible);
-        if let Some(children) = children {
-            for child in children.iter() {
-                let _ = propagate_recursive(
-                    computed_visibility.is_visible_in_hierarchy(),
-                    &mut visibility_query,
-                    &children_query,
-                    *child,
-                    entity,
-                );
+    for (id, vis, parent, children) in &changed {
```

Comment:
Nit: I don't like the use of `id` for a variable name for entities. Please use `entity` like we do in many other places in the code. Also, `vis` -> `visibility`.